### ISSUE TITLE
WebView2 - Selecting text and copying to clipboard

### ISSUE THREAD

**Issue by ajtruckle** [MicrosoftEdge/WebView2Feedback#1474]
With the `CHtmlView` class I was able to select all the text and copy to the clipboard like this:

```
void CChristianLifeMinistryHtmlView::CopyToClipboard()
{
    ExecWB(OLECMDID_COPY, OLECMDEXECOPT_DONTPROMPTUSER, nullptr, nullptr);
}

void CChristianLifeMinistryHtmlView::SelectAll()
{
    ExecWB(OLECMDID_SELECTALL, OLECMDEXECOPT_DONTPROMPTUSER, nullptr, nullptr);
}
```
I am trying to find out how to do the same with the new WebView2 API.

**Comment by ajtruckle** [comment]
I see a similar discussion [here](https://github.com/MicrosoftEdge/WebView2Feedback/issues/564). I can confirm that in the WebView2 control I can use the hotkeys CTRL + A and CTRL + C.

In the linked discussion it says you can programatically do this in VB:

```
Sub Async GetText()
  v = Await wv.ExecuteScriptAsync("document.body.innerText")
End Sub
```

But in my Visual C++ environment I can work out how to execute the above code. I can see "ExecuteScript" method but not the Async version.

**Comment by ajtruckle** [comment]
I have now worked out that this is the right way to get the data:

```
void CWebBrowser::CopyToClipboard()
{
    if (m_pImpl->m_webView != nullptr)
    {
        m_pImpl->m_webView->ExecuteScript(L"document.body.innerText",
            Callback<ICoreWebView2ExecuteScriptCompletedHandler>(
                [](HRESULT error, PCWSTR result) -> HRESULT
                {
                    if (error != S_OK) {
                        ShowFailure(error, L"ExecuteScript failed");
                    }
                    SetClipboardText(result);
                    AfxMessageBox(L"HTML copied to clipboard!", MB_OK);
                    return S_OK;
                }).Get());


    }
}
```

But it does not like my `SetClipboardText(result);` line of code. What is the right way to copy the `result` to the clipboard in `CF_HTML` format?

**Comment by ajtruckle** [comment]
I would personally like to turn this into a feature request please? The above code is not sufficient because it does not return the actual HTML and I have the issue of creating the `CF_HTML` data from the `PCWSTR` variable.

I would be grateful if you can consider adding to WebView2 the functionality that was in original web control that allowed us to programatically select all the content and copy it to the clipboard in HTML format.

At the moment this can be done already with the hotkeys that i specified. So teh functionality is already there. It just needs exposing. At the very least a "CopyToClipboard" function and copied the whole file would be great.

Right now I can't reproduce what the hot keys do programatically.

**Comment by ukandrewc** [comment]
`document.execCommand("copy")`

**Comment by ajtruckle** [comment]
Thanks. Just looking that up here:

https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand

It seems to indicate this command is deprecated and should be avoided. Thoughts?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by ukandrewc** [comment]
True, but it's still there now, new alternative is clipboard API, both here: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Interact_with_the_clipboard

**Comment by ajtruckle** [comment]
Alright, so far I have added the following methods to my browser class:
```

void CWebBrowser::SelectAll()
{
	if (m_pImpl->m_webView != nullptr)
	{
		m_pImpl->m_webView->ExecuteScript(L"document.execCommand(\"SelectAll\")", nullptr);
	}
}


void CWebBrowser::CopyToClipboard2()
{
	if (m_pImpl->m_webView != nullptr)
	{
		m_pImpl->m_webView->ExecuteScript(L"document.execCommand(\"Copy\")", nullptr);
	}
}
```
And I added the following button handlers to my dialog to test:
```
void CMFCTestWebView2Dlg::OnBnClickedButtonSelectAll()
{
	if (m_pWebBrowser != nullptr)
	{
		m_pWebBrowser->SelectAll();
	}
}


void CMFCTestWebView2Dlg::OnBnClickedButtonCopyToClipboard()
{
	if (m_pWebBrowser != nullptr)
	{
		m_pWebBrowser->CopyToClipboard2();
	}
}
```
This does work and appears to be the simplest solution for now. I see the references to the Calendar API that you linked to but I am not sure yet how to implement that alternative.

**Comment by ukandrewc** [comment]
@ajtruckle You seem to be doing this in WebView, not WebView2?

**Comment by ajtruckle** [comment]
No, it is not. Sorry for any confusion. The definition of `m_webView` is:

    wil::com_ptr<ICoreWebView2> m_webView;

**Comment by ukandrewc** [comment]
@ajtruckle Sorry, I'm used to the .Net wrapper where it's `ExecuteScriptAsync`

**Comment by ajtruckle** [comment]
This is in a Visual C++ MFC project and not C# .Net.

![ExecuteScript](https://user-images.githubusercontent.com/7985331/124154359-54a0cb80-da8d-11eb-8e59-9e91f2c2c668.jpg)

**Comment by ukandrewc** [comment]
Yes I realised that after I posted, just saw the name, but should have looked at the framework ;-)

**Comment by ajtruckle** [comment]
😊

I would still like this issue moved into the API request because it is basic functionality and really the API should expose all features that the GUI offers. And the GUI does offer to Select All, and Copy All. Since it is already implemented it makes sense to re use that existing code in the core of the viewer. So simple API commands to select all the content and copy to clipboard would be appreciated by me.

**Comment by ajtruckle** [comment]
I thought I would mention that it seems the Calendar API is currently only supported for reading.

See: [`CoreWebView2PermissionKind Enum`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2permissionkind?view=webview2-dotnet-1.0.864.35)

**Comment by champnic** [comment]
@ajtruckle Thanks for the feature suggestion, I've added this scenario to our backlog. Glad to know you are unblocked for now. Yes, ExecuteScript in Win32 is the same as ExecuteScriptAsync in .NET - both are asynchronous, and are just following different naming conventions for the different frameworks.

**Comment by ajtruckle** [comment]
Thanks for adding to the feature list.

**Comment by ajtruckle** [comment]
Hi @champnic Is there any progress with this one? Thank you.

**Comment by champnic** [comment]
Hey @ajtruckle - No progress on this one.

**Comment by ajtruckle** [comment]
Hi @champnic Alright. You will see a few posts from me (including Stack Overflow).

I have now officially added WebView2 into my app as "beta" but I have some issues. Hence the various tickets and reach outs.

Night.

**Comment by ajtruckle** [comment]
@champnic Hi, is this on the radar yet? Thanks.

**Comment by champnic** [comment]
Hey @ajtruckle - no movement on this one. Given there are workarounds, such as using the Navigator.Clipboard APIs, I don't think we'll be working on this one anytime soon.

**Comment by ajtruckle** [comment]
@champnic 

Hi, I only just spotted this response you made a long time ago on this issue. How do we use these _Navigator.Clipboard APIs_ you refer to? Are you referring to the workaround I already described with `ExecuteScriptAsync`?

**Comment by Stormwalker** [comment]
> How do we use these _Navigator.Clipboard APIs_ you refer to?

I too would like more details on this.  What I would really like to see is the same functionality offered by the WebView2 context menu with respect to clipboard operations so that it can be duplicated in the main menu of my hosting application.  The context menu is able to determine this information at runtime - I don't see why this information cannot be supplied via an API for the host?  Surely clipboard operations are an important part of the UI integration in an hybrid app?

**Comment by champnic** [comment]
Hey @ajtruckle - I believe your `ExecuteScriptAsync` workarounds above used `execCommand` instead of the Clipboard API. The workaround would be similar and use `ExecuteScriptAsync` but you would get whatever info you needed from the webpage and put it into the clipboard using `Navigator.clipboard.write(<data>)` or `.writeText(<text>)`.
https://developer.mozilla.org/en-US/docs/Web/API/Clipboard

@Stormwalker This can definitely be added as an API and that's why we are keeping this item open. We have it prioritized lower on our backlog because there is a workaround using existing web APIs.

**Comment by ajtruckle** [comment]
This is 5 years old. Will it ever fly?

**Comment by Stormwalker** [comment]
Alas, they weren't kidding when they said they'd prioritized it lower :-(
